User Prompt For Benchmark Submission

MSiggel opened the pull request:
Changes made in this pull request:
- moved data frame generation in analyze to separate function
- added user prompt to submit
- information about benchmarks which will be submitted printed prior to the user promt. 


PR Checklist
------------
 - [ ] Added changelog fragment in  `./changelog/` ([more information](https://github.com/bio-phys/MDBenchmark/blob/master/DEVELOPER.rst#using-towncrier))?
 - [ ] Issue raised/referenced?

MSiggel commented:
TODO:
I still need to fix the tests and add new ones for the changes..

mimischi commented on mdbenchmark/analyze.py:
```diff
@@ -33,6 +33,30 @@
 
 plt.switch_backend("agg")
 
+def MakeDataFrameFromBundle(bundle):
```
I would go with a simpler `DataFrameFromBundle`

mimischi commented on mdbenchmark/submit.py:
```diff
@@ -75,39 +85,51 @@ def submit(directory, force_restart):
+        print(df_to_print)
+
+    # here I add the user promt to confirm the submission of the simulations
+    if yes or click.confirm('The above simulations will be submitted. Continue?'):
```
Invert the conditions here. This way we exit early and the biggest part of the code (the currently following lines) will not be indented.

```python
if not yes or not click.confirm('The above simulations will be submitted. Continue?'):
    console.warn("Exiting. No benchmarks submitted.")

grouped_bundles = bundle.categories.groupby("started")
[...]
```

mimischi commented on mdbenchmark/submit.py:
```diff
@@ -75,39 +85,51 @@ def submit(directory, force_restart):
-        "Submitted all benchmarks. Run {} once they are finished to get the results.",
-        "mdbenchmark analyze",
-    )
+    else:
```
This and the following line can be removed, when the if-clause above is inverted (see previous comment).

mimischi commented:
Considering my recent comment, https://github.com/bio-phys/MDBenchmark/issues/80#issuecomment-424269968, we should think about a common format that we want to present to the user. Maybe we could format the "to be submitted" message as a nicely formatted table (not as verbose as the current DataFrame). This we could then reuse for the `mdbenchmark generate` command.

mimischi commented:
@MSiggel I just rebased your PR against the current `develop`. There was a bug preventig Travis-CI from completing.

If you want to do more changes, run `git pull --force`.

MSiggel commented:
@mimischi can you take a look at the output format? Should we consolidate further in one big table to give a single overview of all runs which will be started and chuck all the other information?

mimischi commented on mdbenchmark/analyze.py:
```diff
@@ -33,6 +34,86 @@
+    return df_short
+
+
+def PrintDataFrame(df):
```
What is this for?

mimischi commented on mdbenchmark/generate.py:
```diff
@@ -19,10 +19,12 @@
 from . import console, mdengines, utils
 from .cli import cli
 from .mdengines.utils import write_benchmark
+from .analyze import DataFrameFromBundle, ConsolidateDataFrame, group_consecutives
```
These functions should go into `mdbenchmark.utils`

mimischi commented on mdbenchmark/generate.py:
```diff
@@ -245,6 +247,12 @@ def generate(name, cpu, gpu, module, host, min_nodes, max_nodes, time, skip_vali
                     time=time,
                 )
 
+            bundle = mds.discover(directory)
```
Don't you have the `bundle` above already? `base_directory` defines it for the given loop -- or was that something different? Also maybe we should stick with `dtr` and try to avoid `mds` usage, if we talk about removing it.

mimischi commented on mdbenchmark/generate.py:
```diff
@@ -245,6 +247,12 @@ def generate(name, cpu, gpu, module, host, min_nodes, max_nodes, time, skip_vali
 
+            bundle = mds.discover(directory)
+            df = DataFrameFromBundle(bundle)
+            df_short = ConsolidateDataFrame(df)
```
We need to rewrite this logic, so we first ask the user if they are OK with generating the following benchmarks (printing the summarized DataFrames and throwing a Yes/No prompt at them).

If yes, we actually generate the folders with `write_benchmark`, if no, we cancel and tell them that we did.

To print a summary of the DataFrames at once, you must create an empty list (e.g., `summary_dataframe`) outside of this loop (at the top `for m in module:`), append the `DataFrame` that is returned by `ConsolidateDataFrame(df)` here. After you finish with the `for-loop`, you can easily just print the result of `pd.concat(summary_dataframe)` at once.

MSiggel commented:
@mimischi I fixed yesterdays bugs: runtime is now shown and the table with the summary can display disjoint numbers of nodes in analyze and generate. 
We should definitely think about "beautifying" the table in the output. I added something curde for now.  I think it doesn't stand out at all at the moment because it got much smaller now. Maybe tabulate or some package would be good..if we don't want the added dependency we would have to write something ourselves..
Let me know what you think and after fixing/finalizing this i can adjust all the tests

mimischi commented on mdbenchmark/generate.py:
```diff
@@ -178,7 +180,14 @@ def validate_hosts(ctx, param, host=None):
+@click.option(
+    "-y",
+    "--yes",
+    help="Answer user validation with yes.",
```
Rephrase to "Answer all prompts with yes"

mimischi commented on mdbenchmark/generate.py:
```diff
@@ -201,6 +210,15 @@ def generate(name, cpu, gpu, module, host, min_nodes, max_nodes, time, skip_vali
         console.error("No requested modules available!")
 
+
+    # TODO make dataframe and collect putative setups
```
Is this TODO obsolete now?

mimischi commented on mdbenchmark/generate.py:
```diff
@@ -225,25 +242,42 @@ def generate(name, cpu, gpu, module, host, min_nodes, max_nodes, time, skip_vali
+        print("---------------------------------------------------------")
+
+    if yes:
+        console.info("Generating the above simulation folders.")
```
`"Generating the above benchmarks."`

mimischi commented on mdbenchmark/generate.py:
```diff
@@ -225,25 +242,42 @@ def generate(name, cpu, gpu, module, host, min_nodes, max_nodes, time, skip_vali
+
+    if yes:
+        console.info("Generating the above simulation folders.")
+    elif not click.confirm('The above folders and runs will be submitted. Continue?'):
```
`"The above benchmarks will be generated. Continue?"`

mimischi commented on mdbenchmark/submit.py:
```diff
@@ -59,12 +61,19 @@ def get_batch_command():
     is_flag=True,
 )
-def submit(directory, force_restart):
+@click.option(
```
Missing `help`-text

mimischi commented on mdbenchmark/submit.py:
```diff
@@ -75,38 +84,52 @@ def submit(directory, force_restart):
+
+    # here I add the user promt to confirm the submission of the simulations
+    if yes:
+        console.info("The above simulations will be submitted.")
```
`benchmarks` not `simulations`

mimischi commented on mdbenchmark/submit.py:
```diff
@@ -75,38 +84,52 @@ def submit(directory, force_restart):
+    # here I add the user promt to confirm the submission of the simulations
+    if yes:
+        console.info("The above simulations will be submitted.")
+    elif not click.confirm('The above simulations will be submitted. Continue?'):
```
`benchmarks` not `simulations`

mimischi commented on mdbenchmark/utils.py:
Please provide a source to the StackOverflow answer (not the question). The SO license asks for this.

https://stackoverflow.blog/2009/06/25/attribution-required/

mimischi commented on mdbenchmark/analyze.py:
```diff
@@ -61,22 +61,7 @@ def analyze(directory, plot, ncores, output_name):
-
-    if df.empty:
-        console.error("There is no data for the given path.")
+    df = DataFrameFromBundle(bundle)
```
`DataFrameFromBundle` is not imported and therefore `mdbenchmark analyze` is not working.

MSiggel commented:
@mimischi tabulate is now added and looks pretty nice I think. I added the dependency to setup.py. is there any other place where we need to account for this added dependency?

kain88-de commented:
>  account for this added dependency

conda packages

mimischi commented on mdbenchmark/utils.py:
```diff
@@ -125,3 +131,89 @@ def generate_output_name(extension):
+        node_groups = group_consecutives(df['nodes'].tolist())
+
+        for node_g in node_groups:
+            if len(node_g) is 1:
```
`if len(node_g) == 1`

mimischi commented on mdbenchmark/utils.py:
```diff
@@ -125,3 +131,89 @@ def generate_output_name(extension):
+        for node_g in node_groups:
+            if len(node_g) is 1:
+                 node_print_output.append(node_g[0])
+            elif len(node_g) is not 1:
```
`else`

mimischi commented on setup.py:
```diff
@@ -78,6 +78,7 @@ def get_property(prop, project):
         "matplotlib>=2.2.3,<3.0",  # matplotlib 3.0 does not support Python 2
         "python-Levenshtein",
         "xdg<2",  # xdg 2 does not support Python 2
+        "tabulate"
```
Please pin the version: `tabulate==0.8.2`

MSiggel commented:
@mimischi I updated the tests to accommodate all the new changes in the code concerning the newly formatted outputs. 
We still need to test the prompts and clean up the test code, which I will do in the next commits. Best would be streamlining the design and adding a general file where the text output is generated before being compared to the output of the cli runner.
Should we merge first and take care of the other (minor) changes in a new PR or do it here?

mimischi commented:
I don't think that spending too much time with refactoring tests is a necessity currently. We certainly have to do it, but your changes have not made it more ugly. We'll get back to re-writing those test anyway, as soon as we start with the code refactor.

Just add the tests for input parameters, and we're good to merge this! https://click.palletsprojects.com/en/7.x/testing/#input-streams

mimischi commented on changelog/90.feature.1:
```diff
@@ -0,0 +1 @@
+added user promts to ``mdbenchmark generate`` and ``mdbenchmark submit``
```
```suggestion
Added user prompts to ``mdbenchmark generate`` and ``mdbenchmark submit``.
```

mimischi commented on changelog/90.feature.2:
```diff
@@ -0,0 +1 @@
+added ``--yes`` flag to ``mdbenchmark generate`` and ``mdbenchmark submit`` to bypass user prompt
```
```suggestion
Added ``--yes`` flag to ``mdbenchmark generate`` and ``mdbenchmark submit`` to bypass user prompt.
```

mimischi commented on mdbenchmark/submit.py:
```diff
@@ -60,54 +63,69 @@ def get_batch_command():
+    # Reformat NaN values nicely into question marks.
+    df_to_print = df.replace(np.nan, "?")
+    df_to_print = df.drop(columns=["ns/day", "ncores"])
+    # with pd.option_context("display.max_rows", None):
```
```suggestion

```

Remove the commented code.

mimischi commented on mdbenchmark/submit.py:
```diff
@@ -60,54 +63,69 @@ def get_batch_command():
+    df_to_print = df.replace(np.nan, "?")
+    df_to_print = df.drop(columns=["ns/day", "ncores"])
+    # with pd.option_context("display.max_rows", None):
+    #    print(df_to_print)
```
```suggestion

```

mimischi commented on mdbenchmark/tests/test_generate.py:
```diff
@@ -440,18 +493,31 @@ def test_generate_namd_experimental_warning(cli_runner, monkeypatch, tmpdir):
-            "Creating a total of 5 benchmarks, with a run time of 15 "
-            "minutes each.\nFinished generating all benchmarks.\nYou can "
+        )
+        # cli_runner.invoke(
```
Remove the commented code.
```suggestion

```

mimischi commented on mdbenchmark/tests/test_generate.py:
```diff
@@ -440,18 +493,31 @@ def test_generate_namd_experimental_warning(cli_runner, monkeypatch, tmpdir):
-            "minutes each.\nFinished generating all benchmarks.\nYou can "
+        )
+        # cli_runner.invoke(
+        #    cli.cli, ["analyze", "--output-name=test.csv"]
```
Remove the commented code.
```suggestion

```

mimischi commented on mdbenchmark/tests/test_generate.py:
```diff
@@ -440,18 +493,31 @@ def test_generate_namd_experimental_warning(cli_runner, monkeypatch, tmpdir):
+        )
+        # cli_runner.invoke(
+        #    cli.cli, ["analyze", "--output-name=test.csv"]
+        # )
```
Remove the commented code.
```suggestion

```

mimischi commented on mdbenchmark/tests/test_generate.py:
```diff
@@ -440,18 +493,31 @@ def test_generate_namd_experimental_warning(cli_runner, monkeypatch, tmpdir):
+        # cli_runner.invoke(
+        #    cli.cli, ["analyze", "--output-name=test.csv"]
+        # )
+        # df = pd.read_csv("test.csv", index_col=0)
```
Remove the commented code.
```suggestion

```

mimischi commented on mdbenchmark/tests/test_generate.py:
```diff
@@ -542,3 +608,68 @@ def test_validate_generate_host(ctx_mock):
+            ],
+            input="y\n",
+        )
+        # print(result.output)
```
Remove the commented code.
```suggestion

```

mimischi commented on mdbenchmark/tests/test_generate.py:
```diff
@@ -542,3 +608,68 @@ def test_validate_generate_host(ctx_mock):
+            input="y\n",
+        )
+        # print(result.output)
+        # output = generate_output().format("gromacs/2016")
```
Remove the commented code.
```suggestion

```

mimischi commented on mdbenchmark/tests/test_generate.py:
```diff
@@ -542,3 +608,68 @@ def test_validate_generate_host(ctx_mock):
+        bundle = mds.discover()
+        df = DataFrameFromBundle(bundle)
+        df = ConsolidateDataFrame(df)
+        # test_output = "Simulation Summary:\n" + PrintDataFrame(df, False) + "\n"
```
Remove the commented code.
```suggestion

```

mimischi commented on mdbenchmark/tests/test_utils.py:
```diff
@@ -145,3 +151,62 @@ def dummy(arg):
+
+    expected_output = pd.read_csv(data["analyze-files-gromacs.csv"])
+
+    # TODO: This test fails if we test the dtype. This is weird and I would
```
Is this TODO still relevant?

MSiggel replied:
ah yeah I couldn't figure out why this happens yet. But currently the tests are fine and this is no major problem

mimischi commented on mdbenchmark/utils.py:
```diff
@@ -125,3 +131,97 @@ def generate_output_name(extension):
+        df_short.loc[i] = (key[0], values, df["run time [min]"].iloc[0], key[1], key[2])
+        i += 1
+
+    # df_short.to_csv("output-from-datatframetobundel-short.csv")
```
Remove the commented code.
```suggestion

```
